sklearn/gaussian_process/gpr.py

```diff
@@ -0,0 +1,173 @@
+        self.optimizer = optimizer
+
+    def fit(self, X, y):
+        if self.kernel is None:  # Use an RBF kernel as default
+            self.kernel_ = RBF()
+        else:
+            self.kernel_ = clone(self.kernel)
+
+        X, y = check_X_y(X, y)
+
+        self.X_fit_ = X
+        self.y_fit_ = y
+
+        if self.optimizer == "fmin_l_bfgs_b":
+            # Choose hyperparameters based on maximizing the log-marginal
+            # likelihood using fmin_l_bfgs_b
+            def obj_func(theta):
+                lml, grad = self.log_marginal_likelihood(theta,
+                                                         eval_gradient=True)
+                return -lml, -grad
+            self.theta_, _, _ = fmin_l_bfgs_b(obj_func, self.kernel_.theta,
+                                              bounds=self.kernel_.bounds)
+            self.kernel_.theta = self.theta_
+        elif self.optimizer is None:
+            self.theta_ = self.kernel_.theta
+        else:
+            raise ValueError("Unknown optimizer %s." % self.optimizer)
+
+        # Precompute quantities required for predictions which are independent
+        # of actual query points
+        K = self.kernel_(self.X_fit_)
+        K[np.diag_indices_from(K)] += self.y_err
+        self.L_ = cholesky(K, lower=True)  # Line 2
+        self.alpha_ = cho_solve((self.L_, True), self.y_fit_)  # Line 3
+
+        return self
+
+    def predict(self, X, return_std=False, return_cov=False):
+        if return_std and return_cov:
+            raise RuntimeError(
+                "Not returning standard deviation of predictions when "
+                "returning full covariance.")
+
+        X = check_array(X)
+
+        if not hasattr(self, "X_fit_"):  # Unfitted; predict based on GP prior
+            y_mean = np.zeros(X.shape[0])
+            if return_cov:
+                y_cov = self.kernel(X)
+                return y_mean, y_cov
+            elif return_std:
+                # XXX: Compute y_std more efficiently
+                y_std = np.sqrt(np.diag(self.kernel(X)))
+                return y_mean, y_std
+            else:
+                return y_mean
+        else:  # Predict based on GP posterior
+            K_trans = self.kernel_(X, self.X_fit_)
+            y_mean = K_trans.dot(self.alpha_)  # Line 4 (y_mean = f_star)
+            if return_cov:
+                v = cho_solve((self.L_, True), K_trans.T)  # Line 5
+                y_cov = \
+                    self.kernel_(X) - K_trans.dot(v)  # Line 6
+                return y_mean, y_cov
+            elif return_std:
+                # XXX: Compute y_std more efficiently
+                v = cho_solve((self.L_, True), K_trans.T)  # Line 5
+                y_cov = self.kernel_(X) - K_trans.dot(v)  # Line 6
+                y_std = np.sqrt(np.diag(y_cov))
+                return y_mean, y_std
+            else:
+                return y_mean
+
+    def sample(self, X, n_samples=1, random_state=0):
```

Comment:
That's true; I wasn't aware of the `sample` method in GMMs. I will rename the method to `sample_y`